tests/test_PR045_flatten_operation.py

```diff
@@ -0,0 +1,127 @@
+    array   = awkward1.layout.ListArray64(starts, stops, content)
+    assert awkward1.tolist(array) == [[4, 5, 6], [], [0], [0, 1, 2, 3], [1, 2, 3, 4], [7, 8, 9]]
+    assert flatten(awkward1.tolist(array)) == [4, 5, 6, 0, 0, 1, 2, 3, 1, 2, 3, 4, 7, 8, 9]
+    # Flatten should handle this.
+    assert awkward1.tolist(array.flatten()) == [4, 5, 6, 0, 0, 1, 2, 3, 1, 2, 3, 4, 7, 8, 9]
+
+def test_flatten_list_offset_array():
+    content = awkward1.layout.NumpyArray(numpy.array([0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]))
+    offsets = awkward1.layout.Index64(numpy.array([0, 3, 3, 5, 6, 10], dtype=numpy.int64))
+    array   = awkward1.layout.ListOffsetArray64(offsets, content)
+
+    assert awkward1.tolist(array) == [[0.0, 1.1, 2.2], [], [3.3, 4.4], [5.5], [6.6, 7.7, 8.8, 9.9]]
+    assert awkward1.tolist(array.flatten()) == [0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]
+
+    array2 = array[2:-1]
+    assert awkward1.tolist(array2.flatten()) == [3.3, 4.4, 5.5]
+
+def test_flatten_numpy_array():
+    array = awkward1.layout.NumpyArray(numpy.arange(2*3*5, dtype=numpy.int64).reshape(2, 3, 5))
+    print(awkward1.tolist(array))
+
+    assert awkward1.tolist(array) == [[[ 0,  1,  2,  3,  4], [ 5,  6,  7,  8,  9], [10, 11, 12, 13, 14]],
+                                      [[15, 16, 17, 18, 19], [20, 21, 22, 23, 24], [25, 26, 27, 28, 29]]]
+    assert awkward1.tolist(array.flatten()) == [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9], [10, 11, 12, 13, 14], [15, 16, 17, 18, 19], [20, 21, 22, 23, 24], [25, 26, 27, 28, 29]]
+
+    array2 = awkward1.layout.NumpyArray(numpy.arange(1*6*5, dtype=numpy.int64).reshape(1, 6, 5))
+    print(awkward1.tolist(array2))
+
+    assert awkward1.tolist(array.flatten()) == [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9], [10, 11, 12, 13, 14], [15, 16, 17, 18, 19], [20, 21, 22, 23, 24], [25, 26, 27, 28, 29]]
+
+## def test_flatten_raw_array():
+    ## RawArrayOf<T> is usable only in C++
+
+def test_flatten_record():
+    content = awkward1.layout.NumpyArray(numpy.array([0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]))
+    offsets = awkward1.layout.Index64(numpy.array([0, 3, 3, 5, 6, 10], dtype=numpy.int64))
+    #FIXME: array   = awkward1.layout.Record()
+
+def test_flatten_record_array():
+    content1 = awkward1.layout.NumpyArray(numpy.array([1, 2, 3, 4, 5]))
+    content2 = awkward1.layout.NumpyArray(numpy.array([1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]))
+    offsets = awkward1.layout.Index64(numpy.array([0, 3, 3, 5, 6, 9]))
+    listoffsetarray = awkward1.layout.ListOffsetArray64(offsets, content2)
+    recordarray = awkward1.layout.RecordArray({"one": content1, "two": listoffsetarray})
+
+    assert awkward1.tolist(recordarray) == [{'one': 1, 'two': [1.1, 2.2, 3.3]}, {'one': 2, 'two': []}, {'one': 3, 'two': [4.4, 5.5]}, {'one': 4, 'two': [6.6]}, {'one': 5, 'two': [7.7, 8.8, 9.9]}]
+    #FIXME: assert awkward1.tolist(recordarray.flatten()) == [{'one': [1, 2, 3, 4, 5], 'two': [1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]}]
```

Comment:
You can make a testable RecordArray example by putting the RecordArray inside a ListArray. It's the ListArray that gets flattened, but at least it would test the assertion that ListArray can flatten "any content."

Also, if you want to verify that flattening a RecordArray directly fails as it should, [you can do that with pytest](https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions), and I have examples in some of the other tests.